### PR TITLE
refactor!: rename wasm to smart contract

### PR DESCRIPTION
**Pull request by mversic**
## Description

The fact that we work with wasm smart contracts is a sort of implementation detail.
Let's not specify implementation details in the API.

### Linked issue

Closes #{issue_number} 

### Benefits

If we change from wasm to some other VM, we will have a reduced effect on the API.
If we don't it is implied that we use wasm, so no harm done there

### Checklist

- [ ] I've read `CONTRIBUTING.md`
- [ ] I've used the standard signed-off commit format (or will squash just before merging)
- [ ] All applicable CI checks pass (or I promised to make them pass later)
- [ ] (optional) I've written unit tests for the code changes
- [ ] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

File: client_cli/src/main.rs [hunk 1]
```diff
@@ -108,8 +108,8 @@ enum Subcommand {
     /// The subcommand related to event streaming
     #[clap(subcommand)]
     Events(events::Args),
-    /// The subcommand related to Wasm
-    Wasm(wasm::Args),
+    /// The subcommand related to smart contracts
+    SmartContract(smart_contract::Args),
```
Comment (nxsaken) [inline client_cli/src/main.rs:112]: This changes the subcommand, README needs to be updated

**Review by nxsaken — APPROVED** [review]

**Review by nxsaken — APPROVED** [review]

**Review by nxsaken — APPROVED** [review]

**Review by nxsaken — APPROVED** [review]

File: client/src/client.rs [hunk 2]
```diff
@@ -460,7 +459,9 @@ impl Client {
 
         let mut tx_builder = match instructions.into() {
             Executable::Instructions(instructions) => tx_builder.with_instructions(instructions),
-            Executable::Wasm(wasm) => tx_builder.with_wasm(wasm),
+            Executable::SmartContract(smart_contract) => {
```
Comment (s8sato) [inline client/src/client.rs:462]: Imo what corresponds to smartcontracts is triggers in Iroha, and the execution part of a trigger (and a transaction) can be Wasm or ISIs. So I don't think "smartcontract" is a good candidate to rename "wasm"

Comment (nxsaken) [reply client/src/client.rs:462]: While working on #4850, I noticed that Iroha has both triggers and smart contracts, at least we have two separate macros for defining the entry point. Not sure what the difference is.

Comment (nxsaken) [reply client/src/client.rs:462]: Ok, so smart contracts only have an owner, but triggers also specify the trigger id, owner and event. What happens when a trigger wasm is executed as a smart contract, and vice versa?

Comment (mversic) [reply client/src/client.rs:462]: > While working on #4850, I noticed that Iroha has both triggers and smart contracts, at least we have two separate macros for defining the entry point. Not sure what the difference is.

how I see it.

1. `Transaction::executable` is either a `Vec<Instruction>` or `SmartContract`
You can execute an executable with `client.submit`. This executes the instructions or the smart contract immediately on submition
2. `Action::executable` is either a `Vec<Instruction>` or `SmartContract`
You have to first register a trigger and then invoke it with `InstructionBox::Execute`. The trigger can consist of  `Vec<InstructinoBox`> or a be a smart contract

Comment (mversic) [reply client/src/client.rs:462]: > Ok, so smart contracts only have an owner, but triggers also specify the trigger id, owner and event

no, smart contracts don't have an owner. The transaction they are part of has an owner

Comment (nxsaken) [reply client/src/client.rs:462]: I mean a smart contract entrypoint has an owner field, while a trigger entrypoint has that plus additional fields

**Comment by mversic** [comment]
@s8sato has a good point. I'm closing this

**Pull request closed without merging**